### PR TITLE
Support copy name of active file

### PR DESCRIPTION
**Pull request by iliashkolyar**
Supports #54.
Since VSCode natively supports "Copy Path Of Active File" and "Copy Relative Path Of Active File", I only added "Copy Name Of Active File" which was missing.

### CODE REVIEW

**Comment by iliashkolyar** [comment]
It seems that the test fails as `xclip` isn't installed by default on Linux machines.
Is it possible to add it to the CI running machine or should I disable this test for this OS?

**Comment by sleistner** [comment]
> It seems that the test fails as `xclip` [isn't installed by default on Linux machines](https://github.com/xavi-/node-copy-paste/issues/56#issuecomment-299125937).
> Is it possible to add it to the CI running machine or should I disable this test for this OS?

Would that mean the command would fail on Linux machines?

**Comment by iliashkolyar** [comment]
Yes, on Linux systems that do not have the `xclip` package installed.
I currently suppress the error (if triggers) `Controller`.
Is there a way to show the client the error (for example in the VSCode console) so that he'll install xclip on his machine?

**Comment by sleistner** [comment]
Do you think there's a way to use the VSCode copy/paste system which is used internally?

**Comment by iliashkolyar** [comment]
> Do you think there's a way to use the VSCode copy/paste system which is used internally?

As you can see in the following posts, it seems that vscode are not exposing nor intend to expose their clipboard API:
https://github.com/Microsoft/vscode/issues/10933#issuecomment-242374884
https://github.com/Microsoft/vscode/issues/4972#issuecomment-206288516

When people asked them to support the "Copy File Name", they inclined that [an extension should be used](https://github.com/Microsoft/vscode/issues/47193)

**Comment by sleistner** [comment]
OK, fair enough.

Have you seen the vulnerability message from snyk?

```
sync-exec is an fs.execSync replacement for node <0.12.

Affected versions of this package use tmp directories in an insecure way. 
The file to create will allways return true, regardess if the directory already exists and/or belongs to another user. 
Any user on the server may read the contents of this tmp file and may expose confidential information to an attacker.
```

File: src/controller/CopyFileNameController.ts [hunk 1]
```diff
@@ -14,6 +21,18 @@ export class CopyFileNameController extends BaseCopyController {
         }
 
         const fileName = path.basename(sourcePath);
-        return copyAsync(fileName);
+        return copyAsync(fileName)
+        .catch((error: Error) => {
+            this.handleError(error.message);
```
Comment (sleistner) [inline src/controller/CopyFileNameController.ts]: Error messages are usually handled in `extension.ts`. I would recommend to throw a new Error with updated message.

**Comment by iliashkolyar** [comment]
Replaced `copy-paste` package with a fork that removed the vulnerability (See the [following thread](https://github.com/xavi-/node-copy-paste/pull/64) for more details).

**Comment by sleistner** [comment]
What do you think about the following architecture:

1. Implement `name` as method in `FileItem`
2. CopyFileNameController extends BaseFileController
3. CopyFileNameController#execute returns `Promise<FileItem>`

**Comment by sleistner** [comment]
Maybe something like this.

```ts
export class CopyFileNameCommand extends BaseCommand {

    [...]

    public async execute(uri?: Uri) {
        const sourcePath: string = this.controller.sourcePath;
        if (!sourcePath) {
            return;
        }
        const fileItem = new FileItem(sourcePath);
        return this.controller.execute({ fileItem });
    }

}
```

```ts
export interface IFileController {
    readonly sourcePath: string;
    [...]
}
```

**Comment by iliashkolyar** [comment]
I was thinking that a new interface for different copy operations (incase we'd like to support other copy types) would be better than overloading the fileItem with Copy related fields.

But if you feel that this is a better approach - no problem.

**Comment by sleistner** [comment]
I think it would make sense to keep the api consistent. Also making `name` a getter method of `FileItem` supports the idea of separation of concerns. 
Appreciated if you don't mind changing it.

And thank you very much for contributing, great work!

**Comment by iliashkolyar** [comment]
Up until now, `sourcePath` was a `getter` that was called on demand by the controllers.

If I understood you correctly, here you suggest that `copyFileNameCommand` will access `this.controller.sourcePath`, and that the `sourcePath` parameter will be publicly available on the `BaseFileController`.

What I don't understand is where will it be populated?
You don't have the relevant context when the controller is created.

I can stick with the `util.getSourcePath()` approach so it could be retrieved from both Messages and Controllers.

What do you think?

**Comment by sleistner** [comment]
I meant to extend the `IFileController` interface with 
```ts
readonly sourcePath: string;
```

and change `BaseFileController`
```diff
-    protected get sourcePath(): string {
+    public get sourcePath(): string {
```
 That way you don't need `util` at all.

**Pull request merged**